capa/features/extractors/ts/engine.py

```diff
@@ -0,0 +1,256 @@
+import re
+from typing import List, Tuple, Iterator, Optional
+
+from tree_sitter import Node, Tree, Parser
+
+from capa.features.address import FileOffsetRangeAddress
+from capa.features.extractors.script import LANG_CS, LANG_JS, LANG_TEM, LANG_HTML
+from capa.features.extractors.ts.query import (
+    BINDINGS,
+    QueryBinding,
+    HTMLQueryBinding,
+    ScriptQueryBinding,
+    TemplateQueryBinding,
+)
+from capa.features.extractors.ts.tools import LANGUAGE_TOOLKITS, BaseNamespace, CSharpNamespace, LanguageToolkit
+
+
+class TreeSitterBaseEngine:
+    buf: bytes
+    language: str
+    query: QueryBinding
+    tree: Tree
+
+    def __init__(self, language: str, buf: bytes):
+        self.language = language
+        self.query = BINDINGS[language]
+        self.buf = buf
+        self.tree = self.parse()
+
+    def parse(self) -> Tree:
+        parser = Parser()
+        parser.set_language(self.query.language)
+        return parser.parse(self.buf)
```

Comment:
It can throw type errors (which I believe we prevent with mypy) and a value error if parsing completely fails. Then the parse method will throw a ValueError, so the engine will throw a ValueError etc.: I can handle it in the following way at the Extractor level: